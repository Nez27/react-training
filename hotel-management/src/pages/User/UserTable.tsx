import { useCallback } from 'react';

// Components
import Menus from '@component/Menus';
import Table from '@component/Table';
import Message from '@component/Message';
import Search from '@component/Search';
import SortBy from '@component/SortBy';
import OrderBy from '@component/OrderBy';
import UserRow from './UserRow';
import Pagination from '@component/Pagination';

// Types
import { IUser } from '@type/user';

// Constants
import { ORDERBY_OPTIONS, USER_PAGE } from '@constant/commons';

// Styled
import Direction from '@commonStyle/Direction';
import { StyledOperationTable } from './styled';
import Spinner from '@commonStyle/Spinner';

// Hooks
import { useUsers } from '@hook/users/useUsers';

const UserTable = () => {
  const columnName = [
    'Id',
    'Name',
    'Phone',
    'Is Booked'
  ];
  const {
    isLoading,
    users,
    count
  } = useUsers();

  const renderUserRow = useCallback(
    (user: IUser) => (
      <UserRow
        user={user}
        key={user.id}
      />
    ),
    []
  );

  return (
    <>
      <Direction>
        <StyledOperationTable>
          <OrderBy options={ORDERBY_OPTIONS} />

          <SortBy options={USER_PAGE.SORTBY_OPTIONS} />
          <Search setPlaceHolder="Search by phone..." />
        </StyledOperationTable>

        {isLoading && <Spinner />}

        {users && users.length ? (
          <Menus>
            <Table columns="10% 35% 30% 15% 5%">
              <Table.Header headerColumn={columnName} />
              <Table.Body<IUser> data={users} render={renderUserRow} />
              <Table.Footer>
                <Pagination count={count!}/>
              </Table.Footer>
            </Table>
          </Menus>
        ) : (
          !isLoading && <Message>No data to show here!</Message>
        )}
      </Direction>
    </>
  );
};

export default UserTable;
