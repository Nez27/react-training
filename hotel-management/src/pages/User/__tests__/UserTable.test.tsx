import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { BrowserRouter } from 'react-router-dom';
import UserTable from '@src/pages/User/UserTable.tsx';

// Components

describe('UserTable', () => {
  const queryClient = new QueryClient();

  const wrapper = render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>
        <UserTable />
      </BrowserRouter>
    </QueryClientProvider>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
