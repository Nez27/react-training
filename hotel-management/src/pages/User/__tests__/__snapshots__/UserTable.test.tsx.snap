// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTable Should render correctly 1`] = `
<div
  className="sc-iJCSeZ crZyCc"
  type="vertical"
>
  <div
    className="sc-lmgPJu cuMSiO"
  >
    <div
      className="sc-crzpnZ iahNiP"
    >
      <button
        active={true}
        className="sc-dItHI guVVWw"
        disabled={true}
        onClick={[Function]}
      >
        Ascending
      </button>
      <button
        active={false}
        className="sc-dItHI fIRqtD"
        disabled={false}
        onClick={[Function]}
      >
        Descending
      </button>
    </div>
    <select
      aria-label="Sort"
      className="sc-iqAbyq czAJZn"
      onChange={[Function]}
      value=""
    >
      <option
        value="id"
      >
        Sort by id
      </option>
      <option
        value="name"
      >
        Sort by name
      </option>
      <option
        value="price"
      >
        Sort by price
      </option>
    </select>
    <input
      className="sc-kEqXeH jMVPeG"
      onChange={[Function]}
      placeholder="Search by name..."
    />
  </div>
  <div
    className="sc-giAruI hMNXks"
  />
</div>
`;
